## Final Project Submission

* Student name: Solphine Joseph, Grace Rotich, Mathew Kiprotich, Hilary Simiyu, Clyde Ochieng. 
* Student pace:  full time
* Scheduled project review date/time: 
* Instructor name: Nikita 
* Blog post URL:


# Kings County Housing Analysis with Multiple Linear Regression

## Overview

A real estate agency in Kingsway seeks to determine what are the contributing factors that affect the price of houses to make improvements where necessary. They want to employ an analytical approach rather than sentimental before arriving at a decision. Multilinear regression has been used for this project to understand how various features affect their pricing to better their services.

## Business Problem

In the face of market fluctuations and heightened competition within the real estate sector, our agency is grappling with pricing volatility, which poses significant challenges for our agents in devising effective business strategies. We seek strategic guidance to optimize our purchasing and selling endeavors, prioritizing informed decision-making to identify key areas of focus that promise maximum returns on investment.

### Objectives
* To determine the key factors influencing house prices.
* To develop multilinear regression models to predict house prices based on relevant features.
* To use insights from the regression analysis to optimize pricing strategies for both purchasing and selling properties.


### Hypothesis
* Null Hypothesis - There is no relationship between our independent variables and our dependent variable 

* Alternative Hypothesis - There is a relationship between our independent variables and our dependent variable

### Data Understanding:

In this project, we utilized the King County House Sales dataset, which serves as the foundational dataset for our analysis. It was sourced Kaggle.The dataset encompasses comprehensive information regarding house sales within King County, Washington, USA. It comprises a diverse array of features, including the number of bedrooms, bathrooms, square footage, as well as geographical and pricing details of the properties sold. This dataset is frequently employed in data science and machine learning endeavors, particularly for predictive modeling tasks such as regression analysis aimed at forecasting house prices based on the provided features.

##### King County Housing Data Columns 

The column names contained in column_names.md are:
* `id`: A unique identifier for each house sale.
* `date`: The date when the house was sold.
* `price`: The sale price of the house, serving as the target variable for predictive modeling.
* `bedrooms`, `bathrooms`, `sqft_living`, `sqft_lot`: Numerical features representing the number of bedrooms and bathrooms, as well as the living area and lot area of the house, respectively.
* `floors`: The number of floors in the house.
* `waterfront`, `view`, `condition`, `grade`: Categorical features describing aspects such as waterfront availability, property view, condition, and overall grade assigned to the housing unit.
* `yr_built`, `yr_renovated`: Year of construction and renovation of the house.
* `zipcode`, `lat`, `long`: Geographical features including ZIP code, latitude, and longitude coordinates.
* `sqft_above`, `sqft_basement`, `sqft_living15`, `sqft_lot15`: Additional numerical features providing details about the house's above-ground and basement square footage, as well as living area and lot area of the nearest 15 neighboring houses.

## Data Loading



In [1]:
# Creating a function that loads data and return it in a dataframe
def load_data(file_path):
    house_data = pd.read_csv(file_path)

    #shape
    shape = house_data.shape
    print(f"The dataset contains {shape[0]} houses with {shape[1]} features")
    print()
    
    #Data Types
    data_types = house_data.dtypes
    print("Columns and their data types:")
    for column, dtype in data_types.items():
        print(f"{column}: {dtype}")
    print()

    return house_data


In [2]:
load_data('data/kc_house_data.csv')  # Assuming 'data' folder is in the same directory




NameError: name 'pd' is not defined

The dataset contains 21 columns, each representing a distinct feature, and 21,597 rows, with each row corresponding to a specific house sale entry.

The dataset contains a mix of data types, including integers (int64), floating-point numbers (float64), and objects (strings).  For instance, numerical features such as bedrooms, bathrooms, and sqft_living are represented as integers or floating-point numbers to facilitate mathematical computations, while categorical features like waterfront and view are stored as objects to accommodate text-based categories.

In [ ]:
kings_data = load_data('data/kc_house_data.csv')

The dataset contains 21597 houses with 21 features

Columns and their data types:
id: int64
date: object
price: float64
bedrooms: int64
bathrooms: float64
sqft_living: int64
sqft_lot: int64
floors: float64
waterfront: object
view: object
condition: object
grade: object
sqft_above: int64
sqft_basement: object
yr_built: int64
yr_renovated: float64
zipcode: int64
lat: float64
long: float64
sqft_living15: int64
sqft_lot15: int64



In [ ]:
#create a function that takes in a column and returns the column statistics as a dictionary
def descriptive_analytics(column):
    stats_dict = column.describe().to_dict()
    
    print("Descriptive Statistics for Column '{}':".format(column.name))
    print("The count of the column is:", stats_dict['count'])
    print("The mean of the column is:", stats_dict['mean'])
    print("The standard deviation of the column is:", stats_dict['std'])
    print("The minimum value of the column is:", stats_dict['min'])
    print("The 25th percentile of the column is:", stats_dict['25%'])
    print("The median of the column is:", stats_dict['50%'])
    print("The 75th percentile of the column is:", stats_dict['75%'])
    print("The maximum value of the column is:", stats_dict['max'])

In [ ]:
descriptive_analytics(kings_data['price'])

Descriptive Statistics for Column 'price':
The count of the column is: 21597.0
The mean of the column is: 540296.5735055795
The standard deviation of the column is: 367368.1401013936
The minimum value of the column is: 78000.0
The 25th percentile of the column is: 322000.0
The median of the column is: 450000.0
The 75th percentile of the column is: 645000.0
The maximum value of the column is: 7700000.0


We can see that the maximum price of a house is 7700000 dollars  and the minimum price is 78000 dollars

There are 21597 prices regarding to the houses in the dataset

Average price of a house is 540296.57 dollars

## Data Preperation



In [ ]:
kings_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21597 entries, 0 to 21596
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21597 non-null  int64  
 1   date           21597 non-null  object 
 2   price          21597 non-null  float64
 3   bedrooms       21597 non-null  int64  
 4   bathrooms      21597 non-null  float64
 5   sqft_living    21597 non-null  int64  
 6   sqft_lot       21597 non-null  int64  
 7   floors         21597 non-null  float64
 8   waterfront     19221 non-null  object 
 9   view           21534 non-null  object 
 10  condition      21597 non-null  object 
 11  grade          21597 non-null  object 
 12  sqft_above     21597 non-null  int64  
 13  sqft_basement  21597 non-null  object 
 14  yr_built       21597 non-null  int64  
 15  yr_renovated   17755 non-null  float64
 16  zipcode        21597 non-null  int64  
 17  lat            21597 non-null  float64
 18  long  

In [ ]:
def identify_issues(dataset):
    # Identify missing values as a percentage of the whole dataset
    missing_values = (dataset.isnull().sum())/len(dataset) * 100

    # Identify duplicates
    duplicates = dataset.duplicated().sum()
    
    #return a dictionary 
    return {'duplicates': duplicates,
            'missing values': missing_values.round(2)}   


In [ ]:
identify_issues(kings_data)

{'duplicates': 0,
 'missing values': id                0.00
 date              0.00
 price             0.00
 bedrooms          0.00
 bathrooms         0.00
 sqft_living       0.00
 sqft_lot          0.00
 floors            0.00
 waterfront       11.00
 view              0.29
 condition         0.00
 grade             0.00
 sqft_above        0.00
 sqft_basement     0.00
 yr_built          0.00
 yr_renovated     17.79
 zipcode           0.00
 lat               0.00
 long              0.00
 sqft_living15     0.00
 sqft_lot15        0.00
 dtype: float64}

The examination indicates that there are no duplicate entries within the dataset, ensuring the integrity of the records. However, attention is warranted to address missing values present in certain columns. Specifically, the 'waterfront' feature exhibits 11% of null values, representing a negligible portion of the dataset. Similarly, the 'yr_renovated' feature shows a relatively higher percentage of missing values, accounting for approximately 17.79% of the dataset

##### Before making changes make a copy instead of overwriting data

In [ ]:
house_data_clean = kings_data.copy()

In [ ]:
# Changing the date to date time
house_data_clean['date'] = pd.to_datetime(house_data_clean['date'])

# Extracting only the year from the column Date
house_data_clean.date = house_data_clean['date'].dt.year

# Changing the dates for the year built 
house_data_clean['yr_built'] = pd.to_datetime(house_data_clean['yr_built'],format='%Y').dt.year


The above code converts the 'date' column data to only contain the year the house was sold, for the purpose of analysis we will use only the year since the changes month by month will be minor.

#### Dealing with the missing values

In [ ]:
def missing_values(dataset):
    # drop the rows from views
    dataset.dropna(subset=['view'],inplace=True)

    # Filling the NaN values for waterfront with NO
    dataset.waterfront.fillna('NO',inplace=True)
    
    # Dropping the yr_renovated column 
    dataset.drop('yr_renovated',axis=1,inplace=True)

In [ ]:
missing_values(house_data_clean)

'yr_renovated' has the highest percentage of NaN values 17%. This will be dropped since it will not be used within our model inline with the business problem.

'Waterfront' feature has 11% null values, this was filled with NO on the assumption that these cells were not filled since they lacked waterfronts

For the 'View' column, the null values were dropped by row since the overall percentage impact would be minute

In [ ]:
identify_issues(house_data_clean)

{'duplicates': 2,
 'missing values': id               0.0
 date             0.0
 price            0.0
 bedrooms         0.0
 bathrooms        0.0
 sqft_living      0.0
 sqft_lot         0.0
 floors           0.0
 waterfront       0.0
 view             0.0
 condition        0.0
 grade            0.0
 sqft_above       0.0
 sqft_basement    0.0
 yr_built         0.0
 zipcode          0.0
 lat              0.0
 long             0.0
 sqft_living15    0.0
 sqft_lot15       0.0
 dtype: float64}

In [ ]:
house_data_clean[house_data_clean.duplicated()]

id  date     price  bedrooms  bathrooms  sqft_living  sqft_lot  \
3947   1825069031  2014  550000.0         4       1.75         2410      8447   
20038  8648900110  2014  555000.0         3       2.50         1940      3211   

       floors waterfront  view condition   grade  sqft_above sqft_basement  \
3947      2.0         NO  GOOD      Good  8 Good        2060         350.0   
20038     2.0         NO  NONE   Average  8 Good        1940           0.0   

       yr_built  zipcode      lat     long  sqft_living15  sqft_lot15  
3947       1936    98074  47.6499 -122.088           2520       14789  
20038      2009    98027  47.5644 -122.093           1880        3078